src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/AggregationManager.cs

```diff
@@ -0,0 +1,314 @@
+                {
+                    // this is purely for defensive programming and it should be impossible
+                    // for a developer to observe this exception unless this library has a
+                    // bug in it. Because of that we didn't bother to localize the error message.
+                    throw new InvalidOperationException($"_collectionPeriod must be >= {MinCollectionTimeSecs} sec");
+                }
+
+                DateTime startTime = DateTime.UtcNow;
+                DateTime intervalStartTime = startTime;
+                while (!cancelToken.IsCancellationRequested)
+                {
+                    // intervals end at startTime + X*collectionIntervalSecs. Under normal
+                    // circumstance X increases by 1 each interval, but if the time it
+                    // takes to do collection is very large then we might need to skip
+                    // ahead multiple intervals to catch back up.
+                    //
+                    DateTime now = DateTime.UtcNow;
+                    double secsSinceStart = (now - startTime).TotalSeconds;
+                    double alignUpSecsSinceStart = Math.Ceiling(secsSinceStart / collectionIntervalSecs) *
+                        collectionIntervalSecs;
+                    DateTime nextIntervalStartTime = startTime.AddSeconds(alignUpSecsSinceStart);
+
+                    // The delay timer precision isn't exact. We might have a situation
+                    // where in the previous loop iterations intervalStartTime=20.00,
+                    // nextIntervalStartTime=21.00, the timer was supposed to delay for 1s but
+                    // it exited early so we looped around and DateTime.Now=20.99.
+                    // Aligning up from DateTime.Now would give us 21.00 again so we also need to skip
+                    // forward one time interval
+                    DateTime minNextInterval = intervalStartTime.AddSeconds(collectionIntervalSecs);
+                    if (nextIntervalStartTime <= minNextInterval)
+                    {
+                        nextIntervalStartTime = minNextInterval;
+                    }
+
+                    // pause until the interval is complete
+                    TimeSpan delayTime = nextIntervalStartTime - now;
+                    cancelToken.WaitHandle.WaitOne(delayTime);
+
+                    // don't do collection if timer may not have run to completion
+                    if (cancelToken.IsCancellationRequested)
+                    {
+                        break;
+                    }
+
+                    // collect statistics for the completed interval
+                    _beginCollection(intervalStartTime);
+                    Collect();
+                    _endCollection(intervalStartTime);
+                    intervalStartTime = nextIntervalStartTime;
+                }
+            }
+            catch (Exception e)
+            {
+                _collectionError(e);
```

Comment:
What happens when this fails?  Is there a mechanism for recovery, since Start can only be used once?

Reply:
The recovery mechanism at the moment is to Dispose() the manager and create a new one. We might need a more localized handler to make Exceptions in the user callbacks less problematic as well but I haven't scrutinized that aspect just yet.